server/src/main/java/org/elasticsearch/search/aggregations/bucket/filter/FiltersAggregator.java

```diff
@@ -289,28 +291,66 @@ static class Compatible extends FiltersAggregator {
 
         @Override
         protected LeafBucketCollector getLeafCollector(AggregationExecutionContext aggCtx, LeafBucketCollector sub) throws IOException {
-            if (QueryToFilterAdapter.MatchesNoDocs(filters()) && otherBucketKey == null) {
+            if (QueryToFilterAdapter.matchesNoDocs(filters()) && otherBucketKey == null) {
                 return LeafBucketCollector.NO_OP_COLLECTOR;
             }
-
-            IntPredicate[] docFilters = new IntPredicate[filters().size()];
+            final DocIdSetIterator[] filteredDocIdSetIterators = new DocIdSetIterator[filters().size()];
             for (int filterOrd = 0; filterOrd < filters().size(); filterOrd++) {
-                docFilters[filterOrd] = filters().get(filterOrd).matchingDocIds(aggCtx.getLeafReaderContext());
+                filteredDocIdSetIterators[filterOrd] = filters().get(filterOrd).matchingDocIdSetIterator(aggCtx.getLeafReaderContext());
             }
+
+            final boolean useCompetitiveIterator = (parent == null && otherBucketKey == null);
+            final int maxDocId = aggCtx.getLeafReaderContext().reader().maxDoc();
+
+            // Tracks the last matched doc per filter.
+            final int[] lastMatches = new int[filteredDocIdSetIterators.length];
+            Arrays.fill(lastMatches, -1);
+
+            /*
+             * Even if the competitive iterator is applicable, it may not be used immediately.
+             * We track the last doc id we've processed to effectively switch between the iterator and the direct use
+             * of the filteredDocIdSetIterators in the leaf collector.
+             */
+            final DocId currentDocID = new DocId();
+            currentDocID.set(-1);
+
             return new LeafBucketCollectorBase(sub, null) {
```

Comment:
Maybe return different implementations depending on when competitive iterators can be used?

Reply:
One implementation would keep using Bits and new one would use DocIdSetIterator?

Reply:
I don't think there's any reason to use the Bits implementation, it's equivalent to the one using the DocIdSetIterator? More so, the collector may be used before the competitive iterator kicks in; it may be a testing artifact, a search expert should help clarify this.

Reply:
Right, I think this makes sense. There is not any benefit from keep using Bits based implementation and it doesn't allow us to use competitive iterator.